src/core/packet_builder.c

```diff
@@ -253,12 +253,36 @@ QuicPacketBuilderPrepare(
         //
         BOOLEAN SendDataAllocated = FALSE;
         if (Builder->SendData == NULL) {
+            CXPLAT_ECN_TYPE EcnType =
+                Builder->Path->EcnValidationState == ECN_VALIDATION_CAPABLE ?
+                    CXPLAT_ECN_ECT_0 : CXPLAT_ECN_NON_ECT;
+            if (Builder->Path->EcnValidationState == ECN_VALIDATION_TESTING) {
+                EcnType = CXPLAT_ECN_ECT_0;
+                if (Builder->Path->EcnValidationState == ECN_VALIDATION_TESTING) {
+                    if (Builder->Path->EcnTestingEndingTimeSet) {
+                        if (!CxPlatTimeAtOrBefore64(
+                                CxPlatTimeUs64(), Builder->Path->EcnTestingEndingTime)) {
+                            Builder->Path->EcnValidationState = ECN_VALIDATION_UNKNOWN;
+                        }
+                    } else {
+                        uint64_t TimeNow = CxPlatTimeUs64();
+                        uint32_t ThreePtosInUs =
+                            QuicLossDetectionComputeProbeTimeout(
+                                &Connection->LossDetection,
+                                &Connection->Paths[0],
+                                QUIC_CLOSE_PTO_COUNT);
+                        Builder->Path->EcnTestingEndingTime = TimeNow + ThreePtosInUs;
+                        Builder->Path->EcnTestingEndingTimeSet = TRUE;
+                    }
+                }
+            }
```

Comment:
I think this should either go (or be called from) at the beginning of `QuicSendFlush` or in `QuicPacketBuilderInitialize`.

Reply:
This is somewhat similar to `QuicMtuDiscoveryCheckSearchCompleteTimeout`.

Reply:
If we move it `QuicSendFlush` or `QuicPacketBuilderInitialize`, which is essentially delaying the ECN decision, we will have to update each sent packet's meta data in `QuicSendFlush`, which will be an extra loop. It also conflicts with `CxPlatSendDataAlloc` where it implies the decision should be made when allocating the send context.

Reply:
`QuicSendFlush` or `QuicPacketBuilderInitialize` is before this code path here. I'm suggesting you take the logic you have here, calculate what the ECN info would be for the packets, cache it on the Builder, and then use that value here instead.

Reply:
Done